sql/partitioning/partition_handler.cc

```diff
@@ -639,11 +639,11 @@ int Partition_helper::ph_update_row(const uchar *old_data, uchar *new_data,
     index)
     mysql_update does not set table->next_number_field, so we use
     table->found_next_number_field instead.
-    Also checking that the field is marked in the write set.
+    Also checking that the field is marked in the read set.
   */
   if (m_table->found_next_number_field && new_data == m_table->record[0] &&
       !m_table->s->next_number_keypart &&
-      bitmap_is_set(m_table->write_set,
+      bitmap_is_set(m_table->read_set,
```

Comment:
If we go this way, we will skip `set_auto_increment_if_higher()` call, which is not OK.
The problem is that when binlog is disabled, we are skipping call to `TABLE::mark_columns_per_binlog_row_image()` which is supposed to set table's write/read sets bits. If we skip, then we hit the assert.
On the other hand the assert ASSERT_COLUMN_MARKED_FOR_READ inside `Field_long::val_int()` seems to be too restrictive in this case (that's why commit 74cc73d4619c94ceb1b93725b1bba7b802f2290d introduced `dbug_tmp_use_all_columns / dbug_tmp_restore_column_map` 

My proposal for the fix is:
```
  if (m_table->found_next_number_field && new_data == m_table->record[0] &&
      !m_table->s->next_number_keypart &&
      bitmap_is_set(m_table->write_set,
                    m_table->found_next_number_field->field_index())) {
    my_bitmap_map *old_map = dbug_tmp_use_all_columns(m_table, m_table->read_set);
    set_auto_increment_if_higher();
    dbug_tmp_restore_column_map(m_table->read_set, old_map);
  }
```

Reply:
Yes, but probably `bitmap_is_set` checking also should be removed, because `set_auto_increment_if_higher` and procedures it call don't write to any table fields.

Reply:
I wouldn't touch the `bitmap_is_set` part.

Reply:
The approach to use only dbug_tmp_use_all_columns() looks good to me.